### PR TITLE
feat: Add node-fetch with retry functionality

### LINKED ISSUE #48: Improve error handling when making requests
When api.honeybadger.io fails due to an error (rare, but it can happen, as in #47), webpack compilation fails with a cryptic error:

```
ERROR in HoneybadgerSourceMapPlugin: failed to upload app.bundle.js.map to Honeybadger API: Unexpected token < in JSON at position 0
```

This specific error message resulted from nginx returning a generic 500 error page rather than a JSON response (which is what an error in our API should generate).

Let's add some error handling so that any non-successful response from the server is logged with a warning that the source map upload failed so that the webpack build can continue without erroring out.

### LINKED ISSUE #50: Add config option for API URL
I want to be able to use this library with our internal Honeybadger instance.

Edit: in our client libraries, we usually name this option [`endpoint`](https://www.notion.so/honeybadger/endpoint-4ca023435ff54f949cd7f26be9d56239).

### LINKED ISSUE #281: Find an alternative to `request` npm package
- https://www.npmjs.com/package/request
- Request is deprecated as of Feb 2020, we should seek a replacement.

https://github.com/honeybadger-io/honeybadger-webpack/blob/b110654e24ee55c97ca01b3479432622da2a58d0/package.json#L58

https://github.com/honeybadger-io/honeybadger-webpack/blob/b110654e24ee55c97ca01b3479432622da2a58d0/src/HoneybadgerSourceMapPlugin.js#L2

Perhaps by changing the way HTTP requests are handled could also possibly fix #47

### LINKED ISSUE #285: feat: Should warn user if no assets found
From the Rollbar repo:

https://github.com/thredup/rollbar-sourcemap-webpack-plugin/issues/39

Users would like to see if no assets will be uploaded.

### LINKED ISSUE #286: feat: add retry functionality to sourcemap upload
Based on the rollbar repo:

https://github.com/thredup/rollbar-sourcemap-webpack-plugin/issues/45

https://github.com/thredup/rollbar-sourcemap-webpack-plugin/issues/73

People would like to see retry functionality.

This is also linked to #276 since https://github.com/vercel/fetch-retry claims to solve this issue.

### PR DESCRIPTION
**Pull request by ParamagicDev**
## Status

Ready

## Description

- Add a configurable endpoint
- Add retry functionality to sourcemap post request
- Add a configurable number of retries
- cleanup tests to use `async / await`
- warn users if no assets will be emitted and silent is not true.

## Related Issues

fixes "feat: add retry functionality to sourcemap upload" 
fixes "feat: Should warn user if no assets found"
fixes #281
fixes #50 
fixes #48 , `fetch` separates response errors from network errors. 400+ status codes are treated separately from actual network errors.

Due to not being unable to reproduce even with a 4mb entry file here:

https://github.com/ParamagicDev/rails-shoelace-example/runs/2107323795?check_suite_focus=true

This addresses #276 but I cannot 100% say it will fix it.

## Related PRs

#284 is the precursor to this PR.

branch | PR
------ | ------
chore/package-json-cleanup | [link](https://github.com/honeybadger-io/honeybadger-webpack/pull/284)

## Todos
- [x] Tests
- [x] Documentation
- [x] Changelog Entry (unreleased)

## Steps to Test or Reproduce
Outline the steps to test or reproduce the PR here.

```bash
> git pull --prune
> git checkout feat/switch-to-node-fetch
> npm test
```

### CODE REVIEW

File: test/HoneybadgerSourceMapPlugin.test.js
Comment (joshuap) [inline test/HoneybadgerSourceMapPlugin.test.js:265]: Typo in this sentence?

**Review by joshuap — COMMENTED** [review]
This looks great. Can you update the changelog with a list of changes? They can go under the [unreleased](https://github.com/honeybadger-io/honeybadger-webpack/blob/master/CHANGELOG.md#unreleased) heading for now. As long as there aren't breaking changes, I'd say let's go with an `### Added` subheading (which implies a minor release).

I can do a beta release once this is merged so that we can test it out in our internal projects.

File: test/HoneybadgerSourceMapPlugin.test.js
Comment (ParamagicDev) [reply test/HoneybadgerSourceMapPlugin.test.js:265]: Yup. 100% a typo.

**Review by joshuap — APPROVED** [review]
I adjusted the CHANGELOG to match our preferred format. Looks good! Go ahead and merge this when you're ready and then I'll do a beta release.

**Comment by ParamagicDev** [comment]
All merged in!

**Pull request merged**
